Add a unified resources list screen

bl-nero opened the pull request:
Related issue: #27462

avatus committed:
Update resourceKind method to return slice

avatus committed:
Adds a UnifiedResource watcher to the auth server. This watcher
will watch all the types that are displayed in the web UI and store
them in-memory to allow us to search/filter/query and get multiple
kinds returned at the same time.
[RFD](https://github.com/gravitational/teleport/pull/28162)

avatus committed:
Add UnifiedResourceComponent

avatus committed:
Add unifiedResourceWatcher to auth server

avatus committed:
Init unified resource watcher

avatus committed:
Convert to strings for error output

avatus committed:
Add license

bl-nero committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Add some tests

bl-nero committed:
Merge remote-tracking branch 'origin/master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Fix a race condition and properly close the watcher

avatus committed:
add defer

avatus committed:
Add kubes to watcher

avatus committed:
Add uninit check

avatus committed:
Notify if stale

avatus committed:
Add SAML to watcher

avatus committed:
Update test and fix lint

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Fix license

avatus committed:
Add godocs

avatus committed:
Add Kind to ui structs

avatus committed:
Add Kind to ui structs

avatus committed:
Add FakePaginateParams

avatus committed:
Add FakePaginateParams

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update uses of FakePaginate

avatus committed:
Update uses of FakePaginate

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Update protos

avatus committed:
Update protos

avatus committed:
Add test

avatus committed:
Add test

avatus committed:
Add docs and update tests

avatus committed:
Add docs and update tests

avatus committed:
Add more docs

avatus committed:
Add more docs

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Combine app types in switch

avatus committed:
Combine app types in switch

avatus committed:
Fix tests

avatus committed:
Fix tests

avatus committed:
Add kube access checker

avatus committed:
Add kube access checker

avatus committed:
Add remaining kinds

avatus committed:
Add remaining kinds

bl-nero committed:
Add a Unified Resource card component

bl-nero committed:
WIP: Unified Resource list

bl-nero committed:
Lint fixes

avatus committed:
Review feedback

avatus committed:
Add configurable queuesize to NewWatcher

avatus committed:
Move Get to collector

bl-nero committed:
Review comments

bl-nero committed:
Rename AgentKind and fix lint issues

avatus committed:
make unified resource cache public

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

avatus committed:
Move to common interface

avatus committed:
Use copy as value

avatus committed:
Move saml getter

bl-nero committed:
Review comments

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
WIP: infinite scrolling

bl-nero committed:
Prevent VS Code from showing errors in tsconfig

If tsconfig doesn't have an `outDir` option set, VS Code shows multitude of errors of the following kind:

```
Cannot write file '/Users/bartosz/code/teleport/e/web/teleport/babel.config.js' because it would overwrite input file.
```

bl-nero committed:
Remove unnecessary files

bl-nero committed:
todo

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Cleanup

bl-nero committed:
Add license

zmb3 commented on web/packages/teleport/src/UnifiedResources/Resources.tsx:
```diff
@@ -0,0 +1,86 @@
+  React.useEffect(() => {
+    if (observed.current) {
+      const observer = new IntersectionObserver(entries => {
+        console.log('INTERSECT', entries[0]);
```
Remove log?

zmb3 commented on web/packages/teleport/src/UnifiedResources/Resources.tsx:
```diff
@@ -0,0 +1,86 @@
+    if (observed.current) {
+      const observer = new IntersectionObserver(entries => {
+        console.log('INTERSECT', entries[0]);
+        if (entries[0].isIntersecting) {
```
I've never used an `IntersectionObserver` before - is `entries` guaranteed to be non-empty?

bl-nero replied:
According to MDN, "The list of entries received by the callback includes one entry for each target which reported a change in its intersection status". However, it won't hurt to protect from getting an empty array here, so I'm doing it.

zmb3 commented on web/packages/teleport/src/config.ts:
```diff
@@ -95,6 +95,7 @@ const cfg = {
     clusters: '/web/clusters',
     trustedClusters: '/web/trust',
     audit: '/web/cluster/:clusterId/audit',
+    resources: '/web/cluster/:clusterId/resources',
```
If I try to navigate to this path I just get a 404 error. Am I doing something wrong?

bl-nero replied:
Oh, sorry, forgot to tell you: I decided to hide it behind a `localStorage` setting until we bring the feature set on par with other views. Set `grv_teleport_unified_resources_enabled` in local storage to `true`.

avatus committed:
invert watcher and cache

avatus committed:
use cache instead of watcher

avatus committed:
Fix imports

avatus committed:
Fix lint: check for err

avatus committed:
review round 1

avatus committed:
Update test

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Review, fixing broken tests

avatus committed:
Add copy to get

kimlisa commented on web/packages/teleport/src/services/resources/makeUnifiedResource.ts:
```diff
@@ -0,0 +1,38 @@
+    case 'node':
+      return makeNode(json);
+    case 'windows_desktop':
+      return makeDesktop(json);
```
no kubes?

bl-nero replied:
Wow, thanks for spotting this!

bl-nero replied:
So I added kubes here, but noticed that the API endpoint doesn't return kubes anyway. @avatus , I merged the latest backend changes here and I have one minikube in my configuration. It's visible through a standalone Kubernetes cluster list; however, it's missing from the API response. Could you please look into it? Is it just my setup?

avatus replied:
Not your setup. Found the getter on the backend was not pulling kubes correctly. Will get a fix in

kimlisa commented on web/packages/teleport/src/services/resources/resource.ts:
```diff
@@ -26,6 +30,24 @@ class ResourceService {
       .then(res => makeResourceList<'trusted_cluster'>(res));
   }
 
+  fetchResources(
```
```suggestion
  fetchUnifiedResources(
```

kimlisa commented on web/packages/teleport/src/services/localStorage/localStorage.ts:
```diff
@@ -161,6 +162,12 @@ const storage = {
 
+  areUnifiedResourcesEnabled(): boolean {
+    return (
+      window.localStorage.getItem(KeysEnum.UNIFIED_RESOURCES_ENABLED) === 'true'
```
i'm not sure i'm seeing where this value gets set? (how do we get the true value?)

kimlisa commented on web/packages/teleport/src/config.ts:
```diff
@@ -95,6 +95,7 @@ const cfg = {
     clusters: '/web/clusters',
     trustedClusters: '/web/trust',
     audit: '/web/cluster/:clusterId/audit',
+    resources: '/web/cluster/:clusterId/resources',
```
```suggestion
    unifiedResources: '/web/cluster/:clusterId/resources',
```

kimlisa commented on web/packages/teleport/src/config.ts:
```diff
@@ -151,6 +152,8 @@ const cfg = {
     userStatusPath: '/v1/webapi/user/status',
     passwordTokenPath: '/v1/webapi/users/password/token/:tokenId?',
     changeUserPasswordPath: '/v1/webapi/users/password',
+    resourcesPath:
```
```suggestion
    unifiedResourcesPath:
```

kimlisa commented on web/packages/teleport/src/features.tsx:
```diff
@@ -97,6 +100,30 @@ export class FeatureNodes implements TeleportFeature {
+
+  category = NavigationCategory.Resources;
+
+  hasAccess() {
+    return localStorage.areUnifiedResourcesEnabled();
+  }
```
hrm...i'm not sure how i see local storage being used to set the flag in the future, so if it was a temporary thing i'd remove the local storage changes and manually change flag as you need like this:

```suggestion
  hasAccess() {
    return false // TODO(bl-nero) etc
  }
```

or like this
```
hasAccess(flags: FeatureFlags) {
    return flags.unifiedResources;  // not sure if we ever NOT want to render it though
}
```

and define the flag here: https://github.com/gravitational/teleport/blob/6fb9f081084fda77f1ab0b93834d20efcb843e0c/web/packages/teleport/src/teleportContext.tsx#L95

bl-nero replied:
Actually, I found a better way to do it: instead of using `hasAccess`(), I simply don't register the feature in the first place in `getOSSFeatures()`.

kimlisa commented on web/packages/teleport/src/UnifiedResources/index.ts:
```diff
@@ -0,0 +1,18 @@
+limitations under the License.
+*/
+
+import { Resources } from './Resources';
+export default Resources;
```
idk if it was clear from code base, but we used to "export default" everything, but now we only try to "export default" only for certain things, for this case for use with React.lazy. i try to add comment so we make it clear why it's defaulted here

```suggestion
// export as default for use with React.lazy
export { Resources as default } from './Resources';
```

kimlisa commented on web/packages/teleport/src/UnifiedResources/useResources.ts:
```diff
@@ -0,0 +1,57 @@
+  };
+}
+
+export type State = ReturnType<typeof useResources>;
```
do we have plans for `useResources` hook being used in more than one place? it seems pretty specific to `Resources.tsx` (but idk the full context). If not, we are trying to move away from the `one time use hook` kinda pattern.

We are also moving away from typing our state with `returntype of customhook` kinda pattern too and explicitly define each field (i think the reasoning was it's easier to read)

bl-nero replied:
I agree with the second suggestion. As to the first one, we were following the existing hooks, so yeah, we were just applying an existing pattern. I'd also defend this hook, even if it's a single-use one, since it decouples a part of the logic from the UI. If we cram all these things together, we run into risk of creating another runaway blob of code. WDYT?

kimlisa commented on web/packages/teleport/src/components/hooks/useInfiniteScroll.ts:
```diff
@@ -0,0 +1,116 @@
+    totalCount: 0,
+  });
+
+  const fetch = async () => {
```
```suggestion
  const fetchInit = async () => {
```

bl-nero replied:
I'll go with "fetchInitial", since "fetchInit" sounds as if "init" was a verb, and it's a bit confusing (would it actually fetch, or just initialize fetching?)

avatus committed:
Add CloneAny interface to help with unified resource cloning

avatus committed:
Use cloneany

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Add sort for name and kind

avatus committed:
Add app and saml, fix tests

bl-nero committed:
Review

bl-nero committed:
Lint

avatus committed:
Watch kube server instead of cluster

avatus committed:
Watch kube server instead of cluster

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Get kubecluster from server

avatus committed:
merge master

avatus committed:
update proto file

avatus committed:
Add no access test case

bl-nero commented:
@kimlisa @zmb3 If you don't have any further feedback, may I ask for approval?

avatus committed:
move sorting

avatus committed:
Fix sorting by name

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

avatus committed:
Review feedback round 2

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Remove unneeded bk return

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Keep grpc naming convention

avatus committed:
Add missing docs

avatus committed:
defer log

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

avatus commented on web/packages/teleport/src/UnifiedResources/Resources.tsx:
```diff
@@ -0,0 +1,85 @@
+  const { attempt, fetchedData, fetchMore } = useResources(teleCtx);
+  const observed = React.useRef(null);
+
+  React.useEffect(() => {
+    if (observed.current) {
+      const observer = new IntersectionObserver(entries => {
+        if (entries[0]?.isIntersecting) {
+          fetchMore();
+        }
+      });
+      observer.observe(observed.current);
+      return () => observer.disconnect();
+    }
+  });
```
this is missing a dependency array, which should only require `fetchMore`. Also, this will cause `fetchMore` to run even on the initial load, which we don't want. (These are the times I miss the old class component lifecycle methods). I've seen a pattern before with useEffect where you want it to run on every update to the dependency array but NOT on initial load by using a `isMounted = useRef(false)` type of thing, then setting it to true once you've received the first bit of data or something.

I don't know if I've run into this as issue but just off the top of my head, should we also be checking for `attempt.status !== "processing"` before we 'fetchMore' in this conditional?

bl-nero replied:
And that's a very good catch. This condition may actually fix a bug I saw earlier today.

bl-nero replied:
So about checking for attempt status: the `fetchMore` function already does it internally. I added the missing dependencies, but didn't fix the bug I was talking about ;) I'll fix this one separately.

bl-nero committed:
Fix the useEffect dependency list

avatus committed:
Review feedback

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

avatus committed:
Add ListUnifiedResources gRPC and web endpoints (#29661)

Adds the `ListUnifiedResources` grpc endpoint that returns paginated unified resources and exposes it in the web apiserver

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_resource_watcher' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge remote-tracking branch 'origin/master' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list
